app/uk/gov/hmrc/crdlcache/schedulers/ImportCustomsOfficesListJob.scala

```diff
@@ -0,0 +1,121 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.crdlcache.schedulers
+
+import org.apache.pekko.actor.ActorSystem
+import org.apache.pekko.stream.{ActorAttributes, DelayOverflowStrategy, Supervision}
+import org.apache.pekko.stream.scaladsl.Sink
+import org.mongodb.scala.ClientSession
+import org.quartz.{Job, JobExecutionContext}
+import play.api.Logging
+import uk.gov.hmrc.crdlcache.connectors.DpsConnector
+import uk.gov.hmrc.crdlcache.models.CustomsOfficeListsInstruction.{
+  RecordMissingCustomsOffice,
+  UpsertCustomsOffice
+}
+import uk.gov.hmrc.crdlcache.models.{CustomsOffice, CustomsOfficeListsInstruction}
+import uk.gov.hmrc.crdlcache.repositories.CustomsOfficeListsRepository
+import uk.gov.hmrc.mongo.MongoComponent
+import uk.gov.hmrc.mongo.lock.{LockService, MongoLockRepository}
+import uk.gov.hmrc.mongo.transaction.{TransactionConfiguration, Transactions}
+
+import java.time.{Clock, LocalDate, ZoneOffset}
+import javax.inject.Inject
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration.*
+
+class ImportCustomsOfficesListJob @Inject() (
+  val mongoComponent: MongoComponent,
+  val lockRepository: MongoLockRepository,
+  customsOfficeListsRepository: CustomsOfficeListsRepository,
+  dpsConnector: DpsConnector,
+  clock: Clock
+)(using system: ActorSystem, ec: ExecutionContext)
+  extends Job
+  with LockService
+  with Logging
+  with Transactions {
+
+  given TransactionConfiguration = TransactionConfiguration.strict
+
+  val lockId: String = "import-offices"
+  val ttl: Duration  = 1.hour
+
+  private[schedulers] def processCustomsOffices(
+    session: ClientSession,
+    customsOffices: List[CustomsOffice]
+  ): Future[List[CustomsOfficeListsInstruction]] = {
+    logger.info("Importing customs office lists")
+    customsOfficeListsRepository.fetchCustomsOfficeReferenceNumbers(session).flatMap {
+      existingReferenceNumbers =>
+        val incomingReferenceNumbers = customsOffices
+          .map(_.referenceNumber)
+          .toSet
+        val mergedReferenceNumbers = existingReferenceNumbers.union(incomingReferenceNumbers)
+        val incomingGroupedReferenceNumbers = customsOffices.toSet
+        val instructions                    = List.newBuilder[CustomsOfficeListsInstruction]
+
+        for (referenceNumber <- mergedReferenceNumbers) {
+
+          val hasExistingOffice = existingReferenceNumbers.contains(referenceNumber)
+          val maybeNewOffice = incomingGroupedReferenceNumbers.find(
+            _.referenceNumber == referenceNumber
+          )
```

Comment:
...we can do this here:
```suggestion
          val maybeNewOffice = incomingCustomsOffices
            .get(referenceNumber)
            .flatMap(_.headOption)
```